updater/updater_p3a.cc

```diff
@@ -0,0 +1,85 @@
+/* Copyright (c) 2025 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#include "brave/updater/updater_p3a.h"
+
+#include "base/check.h"
+#include "base/metrics/histogram_macros.h"
+
+namespace brave_updater {
+
+// This pref remembers whether Brave used Omaha 4 or the legacy updater in the
+// last browser launch:
+constexpr char kLastLaunchUsedOmaha4Pref[] =
+    "brave.updater_p3a.last_launch_used_omaha4";
+
+// This pref remembers the version of the browser that was last launched:
+constexpr char kLastLaunchVersionPref[] =
+    "brave.updater_p3a.last_launch_version";
+
+// This pref remembers when the browser was last updated:
+constexpr char kLastUpdateTimePref[] = "brave.updater_p3a.last_update_time";
+
+// This pref remembers whether the last update was done with Omaha 4 or the
+// legacy updater:
+constexpr char kLastUpdateUsedOmaha4Pref[] =
+    "brave.updater_p3a.last_update_used_omaha4";
+
+void RegisterLocalStatePrefs(PrefRegistrySimple* registry) {
+  registry->RegisterStringPref(kLastLaunchVersionPref, "");
+  registry->RegisterBooleanPref(kLastLaunchUsedOmaha4Pref, false);
+  registry->RegisterTimePref(kLastUpdateTimePref, {});
+  registry->RegisterBooleanPref(kLastUpdateUsedOmaha4Pref, false);
+}
+
+void ReportLaunch(base::Time now,
+                  std::string current_version,
+                  bool is_using_omaha4,
+                  PrefService* prefs) {
+  DCHECK(prefs);
+  std::string last_launch_version = prefs->GetString(kLastLaunchVersionPref);
+  prefs->SetString(kLastLaunchVersionPref, current_version);
+
+  bool last_launch_used_omaha4 = prefs->GetBoolean(kLastLaunchUsedOmaha4Pref);
+  prefs->SetBoolean(kLastLaunchUsedOmaha4Pref, is_using_omaha4);
+
+  if (last_launch_version.empty()) {
+    // This is the first launch.
+    return;
+  }
+
+  base::Time last_update_time;
+  bool last_update_used_omaha4;
+
+  if (last_launch_version != current_version) {
+    last_update_time = now;
+    prefs->SetTime(kLastUpdateTimePref, now);
+
+    last_update_used_omaha4 = last_launch_used_omaha4;
+    prefs->SetBoolean(kLastUpdateUsedOmaha4Pref, last_launch_used_omaha4);
```

Comment:
I assume that you duplicate this value because the update system could change at any time after an update. Is that correct? If so, a brief comment could be useful here for future readers.

Reply:
I added a comment in c440ac80f7fef7e6cc54b971ba2ba623ed4a7696.